src/main/java/gumtree/spoon/builder/NodeCreator.java

```diff
@@ -36,28 +60,47 @@ public void scanCtModifiable(CtModifiable m) {
 		String type = MODIFIERS + getClassName(m.getClass().getSimpleName());
 		Tree modifiers = builder.createNode(type, "");
 
-		// We create a virtual node
-		modifiers.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT,
-				new CtVirtualElement(type, m, m.getModifiers(), CtRole.MODIFIER));
-
 		// ensuring an order (instead of hashset)
 		// otherwise some flaky tests in CI
-		Set<ModifierKind> modifiers1 = new TreeSet<>(new Comparator<ModifierKind>() {
-			@Override
-			public int compare(ModifierKind o1, ModifierKind o2) {
-				return o1.name().compareTo(o2.name());
-			}
-		});
-		modifiers1.addAll(m.getModifiers());
+		Set<CtExtendedModifier> modifiers1 = new TreeSet<>(Comparator.comparing(o -> o.getKind().name()));
+		modifiers1.addAll(m.getExtendedModifiers());
+
+		// We create a virtual node
+		var virtualElement = new CtVirtualElement(type, m, m.getModifiers(), CtRole.MODIFIER);
+		modifiers.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, virtualElement);
+
+		List<SourcePosition> positions = new ArrayList<>();
 
-		for (ModifierKind kind : modifiers1) {
-			Tree modifier = builder.createNode("Modifier", kind.toString());
+		for (CtExtendedModifier mod : modifiers1) {
+			Tree modifier = builder.createNode("Modifier", mod.getKind().toString());
 			modifiers.addChild(modifier);
-			// We wrap the modifier (which is not a ctelement)
-			modifier.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, new CtWrapper(kind, m, CtRole.MODIFIER));
+			// We wrap the modifier's kind (which is not a CtElement)
+			var wrapper = new CtWrapper<>(mod.getKind(), m, CtRole.MODIFIER);
+			wrapper.setPosition(mod.getPosition());
+			positions.add(mod.getPosition());
+			modifier.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, wrapper);
 		}
-		builder.addSiblingNode(modifiers);
 
+		//Try to merge source positions of contained elements
+		CompilationUnit cu = null;
+		Integer sourceStart = null;
+		int sourceEnd = 0;
+		for (var position : positions) {
+			if (position instanceof NoSourcePosition) { continue; }
+			if (sourceStart == null || position.getSourceStart() < sourceStart) {
+				sourceStart = position.getSourceStart();
+			}
+			if (position.getSourceEnd() > sourceEnd) {
+				sourceEnd = position.getSourceEnd();
+			}
+			if (cu == null) { cu = position.getCompilationUnit(); }
+		}
+		if (sourceStart != null) {
+			SourcePosition virtualPosition = new SourcePositionImpl(cu, sourceStart, sourceEnd, cu.getLineSeparatorPositions());
+			virtualElement.setPosition(virtualPosition);
+		}
```

Comment:
Preetty sure that this would be accidental. The `CtModifierHandler` creates a hashset. See also https://github.com/INRIA/spoon/issues/4033 for reference. Just don't try to fix that in spoon if you can't spare a week :^)